lib/src/clojure_lsp/kondo.clj

```diff
@@ -54,14 +55,18 @@
 
 (defn db-with-analysis
   "Update `db` with normalized kondo analysis."
-  [db analysis]
-  (update db :analysis merge analysis))
+  [db {:keys [analysis external?]}]
+  (-> db
+      (update :analysis merge analysis)
+      (dep-graph/refresh-analysis (select-keys (:analysis db) (keys analysis))
+                                  analysis
+                                  (not external?))))
```

Comment:
Whenever the analysis is updated, we refresh the dep graph. This is the only place this is done, so to keep the dep graph in sync with the analysis, everything that puts analysis in the db must either call this function or `db-with-results`, which calls this function.

Reply:
I like it how it doesn't have side effects and just return the db with the dep-graph built-in

Reply:
I prototyped the dep-graph maintenance algorithm at the REPL and making it pure made it easier to verify. When I switched to unit tests I could have kept the purity, generating and passing analysis into this function, and checking the results. But I opted for the impure `h/load-code-and-locs` and `f.file-management/did-close` to simulate opening, changing and closing files. I think that's better because it tests that everything integrates with the dep-graph. I guess either way could have worked. Anyway, I'm glad that the core of this is pure... I think Rich's lessons might finally be sinking in. :)